File: module-git/src/main/java/com/timecat/module/git/sgit/activities/delegate/RepoOperationDelegate.java
package com.timecat.module.git.sgit.activities.delegate;

import com.timecat.module.git.android.utils.FsUtils;
import com.timecat.module.git.sgit.activities.RepoDetailActivity;
import com.timecat.module.git.sgit.activities.delegate.actions.AddAllAction;
import com.timecat.module.git.sgit.activities.delegate.actions.AddRemoteAction;
import com.timecat.module.git.sgit.activities.delegate.actions.CherryPickAction;
import com.timecat.module.git.sgit.activities.delegate.actions.CommitAction;
import com.timecat.module.git.sgit.activities.delegate.actions.ConfigAction;
import com.timecat.module.git.sgit.activities.delegate.actions.DeleteAction;
import com.timecat.module.git.sgit.activities.delegate.actions.DiffAction;
import com.timecat.module.git.sgit.activities.delegate.actions.FetchAction;
import com.timecat.module.git.sgit.activities.delegate.actions.MergeAction;
import com.timecat.module.git.sgit.activities.delegate.actions.NewBranchAction;
import com.timecat.module.git.sgit.activities.delegate.actions.NewDirAction;
import com.timecat.module.git.sgit.activities.delegate.actions.NewFileAction;
import com.timecat.module.git.sgit.activities.delegate.actions.PullAction;
import com.timecat.module.git.sgit.activities.delegate.actions.PushAction;
import com.timecat.module.git.sgit.activities.delegate.actions.RawConfigAction;
import com.timecat.module.git.sgit.activities.delegate.actions.RebaseAction;
import com.timecat.module.git.sgit.activities.delegate.actions.RemoveRemoteAction;
import com.timecat.module.git.sgit.activities.delegate.actions.RepoAction;
import com.timecat.module.git.sgit.activities.delegate.actions.ResetAction;
import com.timecat.module.git.sgit.database.models.Repo;
import com.timecat.module.git.sgit.repo.tasks.SheimiAsyncTask.AsyncTaskPostCallback;
import com.timecat.module.git.sgit.repo.tasks.repo.AddToStageTask;
import com.timecat.module.git.sgit.repo.tasks.repo.CheckoutFileTask;
import com.timecat.module.git.sgit.repo.tasks.repo.CheckoutTask;
import com.timecat.module.git.sgit.repo.tasks.repo.DeleteFileFromRepoTask;
import com.timecat.module.git.sgit.repo.tasks.repo.DeleteFileFromRepoTask.DeleteOperationType;
import com.timecat.module.git.sgit.repo.tasks.repo.MergeTask;

import org.eclipse.jgit.lib.Ref;

import java.io.File;
import java.util.ArrayList;

public class RepoOperationDelegate {
    private Repo mRepo;
    private RepoDetailActivity mActivity;
    private ArrayList<RepoAction> mActions = new ArrayList<RepoAction>();

    public RepoOperationDelegate(Repo repo, RepoDetailActivity activity) {
        mRepo = repo;
        mActivity = activity;
        initActions();
    }

    private void initActions() {
        mActions.add(new NewBranchAction(mRepo, mActivity));
        mActions.add(new PullAction(mRepo, mActivity));
        mActions.add(new PushAction(mRepo, mActivity));
        mActions.add(new AddAllAction(mRepo, mActivity));
        mActions.add(new CommitAction(mRepo, mActivity));
        mActions.add(new ResetAction(mRepo, mActivity));
        mActions.add(new MergeAction(mRepo, mActivity));
        mActions.add(new FetchAction(mRepo, mActivity));
        mActions.add(new RebaseAction(mRepo, mActivity));
        mActions.add(new CherryPickAction(mRepo, mActivity));
        mActions.add(new DiffAction(mRepo, mActivity));
        mActions.add(new NewFileAction(mRepo, mActivity));
        mActions.add(new NewDirAction(mRepo, mActivity));
        mActions.add(new AddRemoteAction(mRepo, mActivity));
        mActions.add(new RemoveRemoteAction(mRepo, mActivity));
        mActions.add(new DeleteAction(mRepo, mActivity));
        mActions.add(new RawConfigAction(mRepo, mActivity));
        mActions.add(new ConfigAction(mRepo, mActivity));
    }

    public void executeAction(int key) {
        RepoAction action = mActions.get(key);
        if (action == null)
            return;
        action.execute();
    }

    public void checkoutCommit(final String commitName) {
        CheckoutTask checkoutTask = new CheckoutTask(mRepo, commitName,
                null, new AsyncTaskPostCallback() {
            @Override
            public void onPostExecute(Boolean isSuccess) {
                mActivity.reset(commitName);
            }
        });
        checkoutTask.executeTask();
    }

    public void checkoutCommit(final String commitName, final String branch) {
        CheckoutTask checkoutTask = new CheckoutTask(mRepo, commitName, branch,
                new AsyncTaskPostCallback() {
                    @Override
                    public void onPostExecute(Boolean isSuccess) {
                        mActivity.reset(branch);
                    }
                });
        checkoutTask.executeTask();
    }

    public void mergeBranch(final Ref commit, final String ffModeStr,
                            final boolean autoCommit) {
        MergeTask mergeTask = new MergeTask(mRepo, commit, ffModeStr,
                autoCommit, new AsyncTaskPostCallback() {
            @Override
            public void onPostExecute(Boolean isSuccess) {
                mActivity.reset();
            }
        });
        mergeTask.executeTask();
    }

    public void addToStage(String filepath) {
        String relative = getRelativePath(filepath);
        AddToStageTask addToStageTask = new AddToStageTask(mRepo, relative);
        addToStageTask.executeTask();
    }

    public void checkoutFile(String filepath) {
        String relative = getRelativePath(filepath);
        CheckoutFileTask task = new CheckoutFileTask(mRepo, relative, null);
        task.executeTask();
    }

    public void deleteFileFromRepo(String filepath, DeleteOperationType deleteOperationType) {
        String relative = getRelativePath(filepath);
        DeleteFileFromRepoTask task = new DeleteFileFromRepoTask(mRepo,
                relative, deleteOperationType, new AsyncTaskPostCallback() {
            @Override
            public void onPostExecute(Boolean isSuccess) {
                // TODO Auto-generated method stub
                mActivity.getFilesFragment().reset();
            }
        });
        task.executeTask();
    }

    private String getRelativePath(String filepath) {
        File base = mRepo.getDir();
        String relative = FsUtils.getRelativePath(new File(filepath), base);
        return relative;
    }


}

File: module-git/src/main/java/com/timecat/module/git/sgit/activities/delegate/RepoOperationDelegate.kt
package com.timecat.module.git.sgit.activities.delegate

import com.timecat.module.git.utils.FsUtils.getRelativePath
import com.timecat.module.git.sgit.activities.RepoDetailActivity
import com.timecat.module.git.sgit.activities.delegate.actions.*
import com.timecat.module.git.sgit.database.models.Repo
import com.timecat.module.git.tasks.SheimiAsyncTask.AsyncTaskPostCallback
import com.timecat.module.git.tasks.*
import com.timecat.module.git.tasks.DeleteFileFromRepoTask.DeleteOperationType
import org.eclipse.jgit.lib.Ref
import java.io.File

class RepoOperationDelegate(private val mRepo: Repo, private val mActivity: RepoDetailActivity) {
    private val mActions = ArrayList<RepoAction>()

    init {
        initActions()
    }

    private fun initActions() {
        mActions.add(NewBranchAction(mRepo, mActivity))
        mActions.add(PullAction(mRepo, mActivity))
        mActions.add(PushAction(mRepo, mActivity))
        mActions.add(AddAllAction(mRepo, mActivity))
        mActions.add(CommitAction(mRepo, mActivity))
        mActions.add(ResetAction(mRepo, mActivity))
        mActions.add(MergeAction(mRepo, mActivity))
        mActions.add(FetchAction(mRepo, mActivity))
        mActions.add(RebaseAction(mRepo, mActivity))
        mActions.add(CherryPickAction(mRepo, mActivity))
        mActions.add(DiffAction(mRepo, mActivity))
        mActions.add(NewFileAction(mRepo, mActivity))
        mActions.add(NewDirAction(mRepo, mActivity))
        mActions.add(AddRemoteAction(mRepo, mActivity))
        mActions.add(RemoveRemoteAction(mRepo, mActivity))
        mActions.add(DeleteAction(mRepo, mActivity))
        mActions.add(RawConfigAction(mRepo, mActivity))
        mActions.add(ConfigAction(mRepo, mActivity))
    }

    fun executeAction(key: Int) {
        val action = mActions[key] ?: return
        action.execute()
    }

    fun checkoutCommit(commitName: String) {
        val checkoutTask = CheckoutTask(mRepo, commitName,
            null, object : AsyncTaskPostCallback {
                override fun onPostExecute(isSuccess: Boolean?) {
                    mActivity.reset(commitName)
                }
            })
        checkoutTask.executeTask()
    }

    fun checkoutCommit(commitName: String?, branch: String?) {
        val checkoutTask = CheckoutTask(mRepo, commitName!!, branch!!,
            object : AsyncTaskPostCallback {
                override fun onPostExecute(isSuccess: Boolean?) {
                    mActivity.reset(branch)
                }
            })
        checkoutTask.executeTask()
    }

    fun mergeBranch(
        commit: Ref?, ffModeStr: String?,
        autoCommit: Boolean
    ) {
        val mergeTask = MergeTask(mRepo, commit!!, ffModeStr!!,
            autoCommit, object : AsyncTaskPostCallback {
                override fun onPostExecute(isSuccess: Boolean?) {
                    mActivity.reset()
                }
            })
        mergeTask.executeTask()
    }

    fun addToStage(filepath: String) {
        val relative = getRelativePath(filepath)
        val addToStageTask = AddToStageTask(mRepo, relative)
        addToStageTask.executeTask()
    }

    fun checkoutFile(filepath: String) {
        val relative = getRelativePath(filepath)
        val task = CheckoutFileTask(mRepo, relative)
        task.executeTask()
    }

    fun deleteFileFromRepo(filepath: String, deleteOperationType: DeleteOperationType?) {
        val relative = getRelativePath(filepath)
        val task = DeleteFileFromRepoTask(mRepo,
            relative, deleteOperationType!!, object : AsyncTaskPostCallback {
                override fun onPostExecute(isSuccess: Boolean?) {
                    // TODO Auto-generated method stub
                    mActivity.filesFragment.reset()
                }
            })
        task.executeTask()
    }

    private fun getRelativePath(filepath: String): String {
        val base = mRepo.dir
        return getRelativePath(File(filepath), base)
    }
}